internal/middleware/perfinsights/perfinsights.go

```diff
@@ -0,0 +1,189 @@
+package perfinsights
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/ccoveille/go-safecast"
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/rs/zerolog/log"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+
+	"github.com/authzed/ctxkey"
+	"github.com/authzed/grpcutil"
+
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+)
+
+type APIShapeLabel string
+
+const (
+	ResourceTypeLabel     APIShapeLabel = "resource_type"
+	ResourceRelationLabel APIShapeLabel = "resource_relation"
+	SubjectTypeLabel      APIShapeLabel = "subject_type"
+	SubjectRelationLabel  APIShapeLabel = "subject_relation"
+	NameLabel             APIShapeLabel = "name"
+	FilterLabel           APIShapeLabel = "filter"
+)
+
+var allLabels = []string{
+	string(ResourceTypeLabel),
+	string(ResourceRelationLabel),
+	string(SubjectTypeLabel),
+	string(SubjectRelationLabel),
+	string(NameLabel),
+	string(FilterLabel),
+}
+
+// APIShapeLabels is a map of APIShapeLabel to any value.
+type APIShapeLabels map[APIShapeLabel]any
+
+// NoLabels returns an empty APIShapeLabels map.
+func NoLabels() APIShapeLabels {
+	return APIShapeLabels{}
+}
+
+// APIShapeLatency is the metric that SpiceDB uses to keep track of the latency
+// of API calls, by shape.
+var APIShapeLatency = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace:                   "spicedb",
+	Subsystem:                   "perf_insights",
+	Name:                        "api_shape_latency_seconds",
+	Help:                        "The latency of API calls, by shape",
+	Buckets:                     nil,
+	NativeHistogramBucketFactor: 1.1,
+}, append([]string{"api_kind"}, allLabels...))
+
+// ShapeBuilder is a function that returns a slice of strings representing the shape of the API call.
+// This is used to report the shape of the API call to Prometheus.
+type ShapeBuilder func() APIShapeLabels
+
+// ObserveShapeLatency observes the latency of the API call with the given shape.
+func ObserveShapeLatency(ctx context.Context, methodName string, shape APIShapeLabels, duration time.Duration) {
```

Comment:
please add unit tests, this is complex enough and relevant enough that regressions can be impactful / misleading